webapp-landing/src/main/java/fi/nls/oskari/RssFeedService.java

```diff
@@ -0,0 +1,122 @@
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StreamUtils;
+import java.net.URI;
+import java.nio.charset.Charset;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.List;
+import java.util.Vector;
+
+@Service
+public class RssFeedService implements Runnable {
+
+    private static final String FEED_URL_NOTIFICATIONS = "http://www.maanmittauslaitos.fi/rss/palvelutiedotteet/184";
+    private static final String FEED_URL_NEWS = "http://www.maanmittauslaitos.fi/rss/paikkatietojen-yhteiskaytto";
+    private static int READ_INTERVAL = 60000  * 30; //30 minutes
+    private static int NUM_SHOW_ITEMS = 5;
+
+    private static final Logger logger = Logger.getLogger("RssFeedReader");
+
+    private RssFeedChannel notificationsChannel;
+    private RssFeedChannel newsChannel;
+    private List<RssFeedItem> notifications;
+    private List<RssFeedItem> news;
+    private HttpComponentsClientHttpRequestFactory clientHttpRequestFactory;
+    private ObjectMapper mapper;
+
+    public RssFeedService() {
+        clientHttpRequestFactory = new HttpComponentsClientHttpRequestFactory();
+        mapper = new ObjectMapper();
+        mapper.setDateFormat(new SimpleDateFormat("ddd, dd MMM yyyy HH:mm:SS Z"));
+        new Thread(this).start();
+    }
+
+    public List<RssFeedItem> getNotifications() {
+        return notifications;
+    }
+
+    public List<RssFeedItem> getNews() {
+        return news;
+    }
+
+    private RssFeedChannel readRssFeed(String feedUrl) {
+        try {
+            RssFeedChannel channel = null;
+            ClientHttpRequest req = clientHttpRequestFactory.createRequest(new URI(feedUrl), HttpMethod.GET);
+            ClientHttpResponse resp = req.execute();
+
+            if (resp.getStatusCode() == HttpStatus.OK) {
+                try {
+                    String xml = StreamUtils.copyToString(resp.getBody(), Charset.forName("UTF-8"));
+                    JSONObject jsonObject = XML.toJSONObject(xml);
+                    ObjectMapper objectMapper = new ObjectMapper();
+
+                    RssFeedXmlRoot rssFeedXmlRoot = objectMapper.readValue(jsonObject.toString(), RssFeedXmlRoot.class);
+                    channel = rssFeedXmlRoot.getRssFeed().getChannel();
+                }
+                catch (Exception ex) {
+                    logger.error("Couldn't deserialize RssFeedChannel " +feedUrl, ex);
+                }
+                return channel;
+            }
+            else {
+                logger.warn("RSS feed reader got status " + resp.getRawStatusCode() + "for url " + feedUrl);
+            }
+        }
+        catch (Exception ex) {
+            logger.error("Reading RSS feed failed", ex);
+        }
+        return null;
+    }
+
+    private void updateRssItemLists() {
+        notifications = pickLatestItems(notificationsChannel);
```

Comment:
If an error occured while reading the RSS feed this will replace notifications with an empty list. Perhaps it would be better not to overwrite the current value if an error occurs?

Reply:
Fixed to update notifications and news only if channel is found.